### PR TITLE
Enable .net6.0 build and add net6.0 target

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Fixes #.
[2064](https://github.com/open-telemetry/opentelemetry-dotnet/issues/2064)
## Changes

Please provide a brief description of the changes here.
1) Add .NET6.0 build
2) Add .NET6.0 target to test projects and aspnetcore instrumentation library

todos
1) Replace RC package version to GA versions once available.
2) Remove explicit install for .NET SDKs once 6.0 is available as pre-installed.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: .github/workflows/docfx.yml [hunk 1]
```diff
@@ -15,7 +15,7 @@ jobs:
       uses: actions/checkout@v2
 
     - name: install docfx
-      run: choco install docfx -y
+      run: choco install docfx -y --version=2.58.5
```
Comment (cijothomas) [inline .github/workflows/docfx.yml:18]: can you put a comment why we are using a specific version

File: .github/workflows/code-coverage.yml [hunk 2]
```diff
@@ -26,6 +26,19 @@ jobs:
       with:
         fetch-depth: 0 # fetching all
 
+    - uses: actions/setup-dotnet@v1
```
Comment (eddynaka) [inline .github/workflows/code-coverage.yml:29]: we don't need to install 3.1.x/5.0.x because they are already installed in windows-latest. I always check this: 
- https://github.com/actions/virtual-environments
- https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md

Remember that `windows-latest` can change in the future to windows-2022, so, always have that in mind.

File: .github/workflows/integration.yml [hunk 3]
```diff
@@ -16,9 +16,23 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        version: [netcoreapp3.1,net5.0]
+        version: [netcoreapp3.1,net5.0,net6.0]
     steps:
       - uses: actions/checkout@v2
+
+      - uses: actions/setup-dotnet@v1
```
Comment (eddynaka) [inline .github/workflows/integration.yml:23]: same comment as above, no need for 3.1.x or 5.0.x

File: .github/workflows/integration.yml [hunk 4]
```diff
@@ -27,9 +41,23 @@ jobs:
...
```
Comment (eddynaka) [inline .github/workflows/integration.yml:48]: here as well :)

File: .github/workflows/integration.yml [hunk 5]
```diff
@@ -38,9 +66,23 @@ jobs:
...
```
Comment (eddynaka) [inline .github/workflows/integration.yml:73]: here as well

**Review by reyang — APPROVED** [review]

File: .github/workflows/integration.yml [hunk 6]
```diff
@@ -49,8 +91,22 @@ jobs:
...
```
Comment (eddynaka) [inline .github/workflows/integration.yml:98]: same here.
you can check this:
https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md

File: .github/workflows/linux-ci.yml [hunk 7]
```diff
@@ -16,11 +16,24 @@ jobs:
 
     strategy:
       matrix:
-        version: [netcoreapp3.1,net5.0]
+        version: [netcoreapp3.1,net5.0,net6.0]
 
     steps:
     - uses: actions/checkout@v2
 
+    - uses: actions/setup-dotnet@v1
```
Comment (eddynaka) [inline .github/workflows/linux-ci.yml:24]: same here:
https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md

File: .github/workflows/windows-ci.yml [hunk 8]
```diff
@@ -16,11 +16,24 @@ jobs:
 
     strategy:
       matrix:
-        version: [net461,netcoreapp3.1,net5.0]
+        version: [net461,netcoreapp3.1,net5.0,net6.0]
 
     steps:
     - uses: actions/checkout@v2
 
+    - uses: actions/setup-dotnet@v1
```
Comment (eddynaka) [inline .github/workflows/windows-ci.yml:24]: same here

**Comment by reyang** [comment]
(no need to make the change in this PR)

It seems most of the test projects are using the same target frameworks `netcoreapp3.1;net5.0;net6.0`, might be good to extract the logic to a common place like https://github.com/open-telemetry/opentelemetry-dotnet/blob/02f2594a72503406491ad012edd1f0ef15880cea/docs/Directory.Build.props#L8.

File: test/TestApp.AspNetCore.6.0/Controllers/ChildActivityController.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Mvc;
+using OpenTelemetry;
+
+namespace TestApp.AspNetCore._6._0.Controllers
```
Comment (eddynaka) [inline test/TestApp.AspNetCore.6.0/Controllers/ChildActivityController.cs:22]: one day, we have to figure out how to prevent this code over and over.
we already have the "same" code for 3.1, 5.0, 6.0

File: .github/workflows/integration.yml (hunk 3 above)
Comment (reyang) [reply .github/workflows/integration.yml:23]: `*.tt` files? 🤣

https://docs.microsoft.com/visualstudio/modeling/code-generation-and-t4-text-templates

File: test/TestApp.AspNetCore.6.0/Controllers/ChildActivityController.cs (hunk 9 above)
Comment (reyang) [reply test/TestApp.AspNetCore.6.0/Controllers/ChildActivityController.cs:22]: `*.tt` files? 🤣

https://docs.microsoft.com/visualstudio/modeling/code-generation-and-t4-text-templates

**Comment by vishweshbankwar** [comment]
> (no need to make the change in this PR)
> 
> It seems most of the test projects are using the same target frameworks `netcoreapp3.1;net5.0;net6.0`, might be good to extract the logic to a common place like
> 
> https://github.com/open-telemetry/opentelemetry-dotnet/blob/02f2594a72503406491ad012edd1f0ef15880cea/docs/Directory.Build.props#L8
> 
> .

Yes - I was thinking the same. will look in to creating a separate PR for this.

**Review by eddynaka — APPROVED** [review]

**Review by cijothomas — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetry.Instrumentation.AspNetCore.csproj [hunk 10]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;netstandard2.1;netcoreapp3.1;net5.0</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;netstandard2.1;netcoreapp3.1;net5.0;net6.0</TargetFrameworks>
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetry.Instrumentation.AspNetCore.csproj:3]: Just curious, why was this needed?

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs [hunk 11]
```diff
@@ -52,7 +52,7 @@ public class AspNetCoreInstrumentationOptions
         /// </remarks>
         public bool RecordException { get; set; }
 
-#if NETSTANDARD2_1 || NETCOREAPP3_1 || NET5_0
+#if NETSTANDARD2_1 || NETCOREAPP3_1 || NET5_0_OR_GREATER
```
Comment (bruno-garcia) [inline src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs:55]: Can be only:

```
#if NETSTANDARD2_1 || NETCOREAPP3_1_OR_GREATER
```

Comment (mic-max) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs:55]: The reason why we probably used `#if NETSTANDARD2_1 || NETCOREAPP3_1 || NET5_0` in the first place
https://docs.microsoft.com/en-us/dotnet/core/compatibility/sdk/5.0/netcoreapp3_1-preprocessor-symbol-not-defined

Comment (bruno-garcia) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs:55]: That affects .net 5 SDK, right? So now on NET6 it should work if I undersand?

File: src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetry.Instrumentation.AspNetCore.csproj (hunk 10 above)
Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetry.Instrumentation.AspNetCore.csproj:3]: As we are using <FrameworkReference Include="Microsoft.AspNetCore.App" />, this is needed.
IIRC, if we do not add this, net5.0 targeted instrumentation library will be used for net6.0 app (should still work as we don't have any NET6 framework specific code).

Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetry.Instrumentation.AspNetCore.csproj:3]: I don't _think_ it is needed until we want to use some piece of the .NET 6 API.

If the project is like this...

```csharp
<TargetFrameworks>netstandard2.0;netstandard2.1;netcoreapp3.1;net5.0</TargetFrameworks>

  <ItemGroup Condition="'$(TargetFramework)' == 'net5.0'">
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
  </ItemGroup>
```

Users on .NET 6 will still be able to use it.

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetry.Instrumentation.AspNetCore.csproj:3]: Agreed - then do we remove 5.0 as well?. I have not tested but I think having only 3.1 will also work for 3.1+ apps then?.
But if we ever use version specific code, we will have to do add it again.
@cijothomas - thoughts?

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/OpenTelemetry.Instrumentation.AspNetCore.csproj:3]: We don't use any .NET 6 specific feature today, but Asp.Net core 6.0 has introduced ActivitySource usage. If we want to leverage them, we either need .net6.0 target. or we use reflection to find if user is running Asp.Net Core 6.0 or newer.
(@vishweshbankwar please create a separate issue to discuss this further)

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs (hunk 11 above)
Comment (mic-max) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationOptions.cs:55]: Yes, it looks like it from the [.NET 5 and later](https://github.com/dotnet/designs/blob/main/accepted/2020/or-greater-defines/or-greater-defines.md#design) bullet here
